src/kinetics/Falloff.cpp

```diff
@@ -12,23 +12,147 @@
 #include "cantera/base/AnyMap.h"
 #include "cantera/kinetics/Falloff.h"
 
+
 namespace Cantera
 {
 
 void Falloff::init(const vector_fp& c)
+{
+    setData(c);
+}
+
+void Falloff::setLowRate(const ArrheniusRate& low)
+{
+    if (low.preExponentialFactor() < 0 && !allow_negative_pre_exponential_factor) {
+        throw CanteraError("Falloff::setLowRate",
+            "Detected negative pre-exponential factor (A={}).\n"
+            "Enable 'allow_negative_pre_exponential_factor' to suppress "
+            "this message.", low.preExponentialFactor());
+    } else if (std::isnan(m_highRate.preExponentialFactor())) {
+        // pass
+    } else if (m_highRate.preExponentialFactor() * low.preExponentialFactor() < 0.) {
+        throw CanteraError("Falloff::setLowRate",
+            "Detected inconsistent rate definitions;\nhigh and low "
+            "rate pre-exponential factors must have the same sign.");
+    }
+    m_lowRate = low;
+}
+
+void Falloff::setHighRate(const ArrheniusRate& high)
+{
+    if (high.preExponentialFactor() < 0 && !allow_negative_pre_exponential_factor) {
+        throw CanteraError("Falloff::setHighRate",
+            "Detected negative pre-exponential factor (A={}).\n"
+            "Enable 'allow_negative_pre_exponential_factor' to suppress "
+            "this message.", high.preExponentialFactor());
+    } else if (std::isnan(m_lowRate.preExponentialFactor())) {
+        // pass
+    } else if (m_lowRate.preExponentialFactor() * high.preExponentialFactor() < 0.) {
+        throw CanteraError("Falloff::setHighRate",
+            "Detected inconsistent rate definitions;\nhigh and low "
+            "rate pre-exponential factors must have the same sign.");
+    }
+    m_highRate = high;
+}
+
+void Falloff::setData(const vector_fp& c)
 {
     if (c.size() != 0) {
-        throw CanteraError("Falloff::init",
+        throw CanteraError("Falloff::setData",
             "Incorrect number of parameters. 0 required. Received {}.",
             c.size());
     }
 }
 
-void Troe::init(const vector_fp& c)
+void Falloff::getData(vector_fp& c) const
+{
+    c.clear();
+}
+
+void Falloff::setParameters(const AnyMap& node, const UnitsVector& rate_units)
+{
+    if (node.empty()) {
+        return;
+    }
+
+    allow_negative_pre_exponential_factor = node.getBool("negative-A", false);
+    if (node["type"] == "chemically-activated") {
+        m_chemicallyActivated = true;
+    }
+
+    UnitsVector low_rate_units = rate_units;
+    UnitsVector high_rate_units = rate_units;
+    if (rate_units.size()) {
+        high_rate_units.pop_back();
+        if (m_chemicallyActivated) {
+            low_rate_units.pop_back();
+            high_rate_units.pop_back();
+        }
```

Comment:
While what's highlighted indeed implies what you're saying, what is implemented should be correct: in effect,  `high_rate_units.pop_back()` needs to be called twice (the previous call is on the line above what's shown here), so this is independent of reaction order. You can verify that this with the following:
```
In [1]: import cantera as ct
   ...: gas = ct.Solution('test/data/chemically-activated-reaction.xml')
   ...: gas.reaction(0).reaction_type
Out[1]: 'chemically-activated-legacy'

In [2]: data = gas.reaction(0).input_data
   ...: rxn = ct.Reaction().from_dict(data, gas)
   ...: rxn # illustrates that this is (again) recognized with the correct type
Out[2]: <ChemicallyActivatedReaction: ch3 + oh (+M) <=> ch2o + h2 (+M)>

In [3]: gas.forward_rates_of_progress
Out[3]: array([943.07718202, 943.07718202])
```
So what's implemented indeed replicates the old behavior. I believe the difference here is that it's necessary to peel off the third-body collision partner no matter what.

As an aside, you see why I implemented some of the 'empty' constructors here (`ct.Reaction().from_dict(...)`). The same applies for `ct.ReactionRate().from_dict()`, which should provide some context for my other responses.

Reply:
Sorry, GitHub didn't show quite the right lines -- by "second" `pop_back`, I was referring to the second `pop_back` on `high_rate_units`. The two values that were being removed were first the standard concentration units (corresponding to the third body, I suppose) and then the units corresponding to the last reactant. If that reactant had a stoichiometric coefficient other than one, then this would result in the units of the two rate coefficients differing by more than factor of the standard concentration, which isn't consistent with the rest of the formulation.

In any case, the new implementation looks correct to me, so I guess this is a moot point.

Reply:
Yes, you are correct - I read the "second" incorrectly due to the GitHub formatting, and the assessment was indeed spot-on. But as you saw, the new `UnitStack` sidesteps this issue.